File: depth_first_search/1443_Minimum_Time_to_Collect_All_Apples_in_a_Tree/1443_Minimum_Time_to_Collect_All_Apples_in_a_Tree.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]],
                hasApple: List[bool]) -> int:
        # 初始化路径
        maps = collections.defaultdict(list)
        for e in edges:
            maps[e[0]].append(e[1])

        def dfs(i):
            selfOrChildHasApple = hasApple[i]
            for nex in maps[i]:
                selfOrChildHasApple |= dfs(nex)
            if not selfOrChildHasApple:
                # 从字典中移除自身或子树都没有苹果的节点
                del maps[i]
            return selfOrChildHasApple

        dfs(0)
        # 字典个数即为最终有效节点的个数
        # 但是有可能有效节点为0, 所以需要max一下
        return max(0, 2 * (len(maps) - 1))

File: depth_first_search/1443_Minimum_Time_to_Collect_All_Apples_in_a_Tree/1443_Minimum_Time_to_Collect_All_Apples_in_a_Tree.go
package main

func minTime(n int, edges [][]int, hasApple []bool) int {
	graph := make(map[int][]int)
	for _, v := range edges {
		a := v[0]
		b := v[1]
		if _, ok := graph[a]; !ok {
			graph[a] = []int{}
		}
		graph[a] = append(graph[a], b)
		if _, ok := graph[b]; !ok {
			graph[b] = []int{}
		}
		graph[b] = append(graph[b], a)
	}
	set := make(map[int]bool)
	set[0] = true
	return dfs(0, graph, set, hasApple)
}

func dfs(cur int, graph map[int][]int, set map[int]bool, has []bool) int {
	set[cur] = true
	res := 0
	for _, v := range graph[cur] {
		if !set[v] {
			res += dfs(v, graph, set, has)
		}
	}
	if cur == 0 {
		return res
	}
	if res > 0 || has[cur] {
		res += 2
	}
	return res
}
